### ISSUE TITLE
Dramatic reduction in CPU utilization on upgrading nd4j.version to 0.5.0 from 0.4.0

### ISSUE THREAD

**Issue by ChessPoker** [deeplearning4j/nd4j#1241]
On ND4J 0.4.0 typical CPU usage is ~3700%.
On ND4j 0.5.0 typical CPU usage is ~600%.

Epoch elapsed times do not correspond, however. The first epoch is 100% faster on 0.5.0, but subsequent epochs are ~15% slower on 0.5.0.

Below are two screenshots of hotspots. First is on 0.4.0. 
![zero four zero](https://cloud.githubusercontent.com/assets/8595703/18027020/e21380ce-6c0c-11e6-8e8a-377375cf5ea2.png)

Next image is hotspots on 0.5.0.

![zero five zero](https://cloud.githubusercontent.com/assets/8595703/18027024/08ae171c-6c0d-11e6-950c-acd4ffd681ed.png)

OS is Linux Mint 17.3 -> ldd (Ubuntu EGLIBC 2.19-0ubuntu6.7) 2.19

CPU is **dual processor**: 2 x Intel Xeon E5-2687Wv3, 3.1 GHz (10-Core, HT, 25MB Cache, 160W) 22nm.

**Comment by raver119** [comment]
Can i have your NN configuration please?

**Comment by ChessPoker** [comment]
MultiLayerNetwork net;
final int[] LSTMLayers = new int[]{6};
NeuralNetConfiguration.ListBuilder list = new NeuralNetConfiguration.Builder()                    .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT).iterations(1)
                    .learningRate(0.02)
                    .seed(192125)
                    .weightInit(WeightInit.XAVIER)
                    .updater(Updater.ADAM).adamMeanDecay(0.99).adamVarDecay(0.9999)
                    .list();
            int layerIdx = 0, nIn = iter.inputColumns(), nOut = iter.totalOutcomes();
            for (; layerIdx < LSTMLayers.length; layerIdx++) {
                list = list.layer(layerIdx, new GravesLSTM.Builder().nIn(nIn).nOut(LSTMLayers[layerIdx])
                        .activation("tanh").build());
                nIn = LSTMLayers[layerIdx];
            }
            list.layer(layerIdx++, new RnnOutputLayer.Builder(LossFunctions.LossFunction.MCXENT).activation("softmax")
                    .nIn(nIn).nOut(nOut).build());
MultiLayerConfiguration conf = list.pretrain(false).backprop(true).build();

**Comment by ChessPoker** [comment]
GravesLSTMCharModellingExample in examples is ~4x slower on 0.5.0. It's about 1.8 seconds per iteration on 0.4.0. On 0.5.0 it's over 7 seconds.

CPU utilization was about 3700% in both cases. So this instance is different than above in that the performance is much worse overall, but CPU utilization unchanged between 0.4.0 and 0.5.0.

Below is 0.4.0 hotspots followed by 0.5.0.
![zero four zero charmodellingexample](https://cloud.githubusercontent.com/assets/8595703/18027422/155eeb0c-6c18-11e6-9b48-74b31d2894eb.png)

![zero five zero charmodellingexample](https://cloud.githubusercontent.com/assets/8595703/18027429/1ea98122-6c18-11e6-95e3-0936b98d6e35.png)

**Comment by saudet** [comment]
What happens with `export OMP_NUM_THREADS=1` or 2?

**Comment by ChessPoker** [comment]
===On 0.4.0 ===
With OMP_NUM_THREADS=1 performance is tremendously improved (around x4 faster) vs no setting for OMP_NUM_THREADS.

With OMP_NUM_THREADS=2 I don't notice a difference from OMP_NUM_THREADS=1, but the elapsed times are noisy. It's possible there's a slight difference.

===On 0.5.0 ===
With OMP_NUM_THREADS=1 performance is the same as 0.4.0 and OMP_NUM_THREADS=1. That is, very good performance.
With OMP_NUM_THREADS=2  no noticeable difference.
With OMP_NUM_THREADS=4 no noticeable difference.

So wow! I thought dual 10-core Xeons were helping me. Turns out I just need one core.

**Comment by raver119** [comment]
Nah, they still can help you, if you'll use ParallelWrapper and train your model in data-parallel mode.

**Comment by raver119** [comment]
This issue should be fixed now.

**Comment by ChessPoker** [comment]
Is it fixed in 0.5.1-SNAPSHOT? I'm on that version now and it's still requiring OMP_NUM_THREADS=1 to get good performance and cpu utilization is lowish.

**Comment by raver119** [comment]
https://github.com/deeplearning4j/libnd4j/issues/287

Take a look here, that's mostly about cpu affinity now.

**Comment by raver119** [comment]
Lol it's you. Nvm then.
